pa_matchbatch.sh

```diff
@@ -0,0 +1,149 @@
+#!/usr/bin/env bash
+set -e
+unset CDPATH
+
+# where to find the agent; can be overridden
+puppet_agent_repo="git@github.com:puppetlabs/puppet-agent.git"
+PUPPET_AGENT_URL=${PUPPET_AGENT_URL:-${puppet_agent_repo}}
+
+# where to find the ticketmatch.rb script; can be overridden
+TICKETMATCH_PATH=${TICKETMATCH_PATH:-$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )}
+
+# shorthand
+pushd() {
+	builtin pushd "$@" > /dev/null
+}
+
+popd() {
+	builtin popd "$@" > /dev/null
+}
+
+# get rev hashes in the form [to_rev]|[url]
+getComponentRevMap() {
+	# looking for components not pinned to a 'refs/tags' element, and of those, filtering out (keeping) the ones owned by puppetlabs
+	for componentName in $(for componentFile in $(grep -lv refs/tags configs/components/*.json); do grep -l puppetlabs ${componentFile}; done); do 
+		ruby -rjson -e'j = JSON.parse(STDIN.read); printf(" %s|%s", j["ref"], j["url"])' < ${componentName}
+	done
+}
+
+# TODO: more robust version matches?
+getFixVerFor() {
+	case ${1} in
+		puppet-resource_api)
+			git log -1 --no-merges --oneline --grep='Release prep' | sed -e "s/.*prep for v\(.*$\)/\1/"
+		;;
+		*)
+			git log -1 --no-merges --oneline -E --grep='\(packaging\) .* version' | sed -e "s/.*'\(.*\)'.*/\1/"
+		;;
+	esac
+}
+
+getJiraProjectIdFor() {
+	case "${1}" in
+		facter) echo FACT
+		;;
+		hiera) echo HI
+		;;
+		leatherman) echo LTH
+		;;
+		puppet) echo PUP
+		;;
+		pxp-agent) echo PCP
+		;;
+		puppet-agent) echo PA
+		;;
+		cpp-pcp-client) echo PCP
+		;;
+		libwhereami) echo FACT
+		;;
+		marionette-collective) echo MCO
+		;;
+		puppet-resource_api) echo PDK
+		;;
+		cpp-hocon) echo HC
+		;;
+		nssm) echo PA  # this is on purpose
+		;;
+		*) (>&2 echo "Error: need to add JIRA project mapping for '${1}'.")
+			exit 1
+		;;
+	esac
+}
+
+getJiraFixedInFor() {
+	case "${1}" in
+		facter) echo FACT
+		;;
+		hiera) echo HI
+		;;
+		leatherman) echo LTH
+		;;
+		puppet) echo PUP
+		;;
+		pxp-agent) echo pxp-agent
+		;;
+		puppet-agent) echo puppet-agent
+		;;
+		cpp-pcp-client) echo cpp-pcp-client
+		;;
+		libwhereami) echo whereami  # potential headache
+		;;
+		marionette-collective) echo MCO
+		;;
+		puppet-resource_api) echo PDK
+		;;
+		cpp-hocon) echo HC
+		;;
+		nssm) echo puppet-agent  # this is on purpose
+		;;
+		*) (>&2 echo "Error: need to add JIRA fixed-in version mapping for '${1}'.")
+			exit 1
+		;;
+	esac
+}
+
+# main
+puppetAgentBaseRev=${1}
+[[ -z ${puppetAgentBaseRev} ]] && { echo "Error: must specify the puppet-agent revision to start from."; exit 1; }
+
+baseDir=${WORKSPACE:-${PWD}}
+[[ -d ${baseDir} ]] || { echo "Error: '${baseDir}' -- no such directory."; exit 1; }
+cd ${baseDir}
+[[ $(ls -A | wc -l) -eq 0 ]] || { echo "Error: '${PWD}' is not empty."; exit 1; }
+
+# get desired revision of puppet-agent in place
+echo "Note: cloning puppet-agent..."
+git clone --quiet ${PUPPET_AGENT_URL}
```

Comment:
If you are going to be using the read_repo_version thing I linked in an earlier comment, then you might want to specify the private forks as part of the special cases (or maybe parse out the "-private" part altogether as a first-pass filter in the routine).